src/Utilities/Compiler/Extensions/SourceTextExtensions.cs

```diff
@@ -0,0 +1,47 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.IO;
+using System.Text;
+using Microsoft.CodeAnalysis.Text;
+
+namespace Analyzer.Utilities.Extensions
+{
+    /// <summary>
+    /// Class that contains extensions to <see cref="SourceText"/>.
+    /// </summary>
+    internal static partial class SourceTextExtensions
+    {
+        /// <summary>
+        /// Reads the <paramref name="text"/> contents into a stream and returns the result of calling the
+        /// <paramref name="parser"/> function on that stream.
+        /// </summary>
+        /// <typeparam name="T">Type to deserialize from the <paramref name="text"/>.</typeparam>
+        /// <param name="text">Abstraction for an additional file's contents.</param>
+        /// <param name="parser">Function that will parse <paramref name="text"/> into <typeparamref name="T"/>.</param>
+        /// <returns>Output from <paramref name="parser"/>.</returns>
+        public static T Parse<T>(this SourceText text, Func<StreamReader, T> parser)
```

Comment:
❕  Use `AnalysisContext.TryGetValue` instead. This method can be removed.

Reply:
Wouldn't I still need code to read/parse SourceText?

Reply:
That's what [`TryGetValue`](https://github.com/dotnet/roslyn/blob/1309e6b088a884104057b7614389493d436b3aee/src/Compilers/Core/Portable/DiagnosticAnalyzer/DiagnosticAnalysisContext.cs#L226-L236) does.

Reply:
Haha, `TryGetValue` calls `SourceText.[Lines|Write]`? In my code, `TryGetValue` takes `SourceText` as the "key" and either `_xmlDictionaryProvider` or `_dicDictionaryProvider` as the `SourceTextValueProvider`. These providers read/parse the underlying text from `SourceText` with this method and map it to a `CodeAnalysisDictionary`. It's this mapped value that `TryGetValue` returns either by cache or calculation.

Reply:
@sharwell is correct. If multiple analyzers end up requiring converting the same source text instance to dictionary value, using `AnalysisContext.TryGetValue` will ensure the compute is shared.

Reply:
Ah, so reading the code, this seems like the core helper method used by the `AnalysisContext.TryGetValue` callback, correct? If so, this seems reasonable to me.

Reply:
Right